### ISSUE TITLE
size regression due to host view factories generated for non-entryComponents

### ISSUE THREAD

**Issue by IgorMinar** [angular/angular#19857]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior
in v4 the AOT compiler generated host view factories only for entryComponents in v5.0.0.rc2 we generate them even for components that are not marked as entryComponents. 

Example in aio is SearchResultsComponent, which is not marked as entryComponent. When upgrading from 5.0.0-beta.3, I noticed that we now started generating `View_SearchResultsComponent_Host_0` for this component. And we do it in a way that tightly couples it with the rest of the component factories, so it can't be eliminated.

Relevant snippets of the generated code:

```js
function View_SearchResultsComponent_Host_0(_l) {
  return viewDef(0, [ (_l()(), elementDef(0, 0, null, null, 1, "aio-search-results", [], null, null, null, View_SearchResultsComponent_0, RenderType_SearchResultsComponent)), directiveDef(1, 573440, null, 0, search_results_component_SearchResultsComponent, [], null, null) ], null, null);
}

styles_FileNotFoundSearchComponent = (createComponentFactory("aio-search-results", search_results_component_SearchResultsComponent, View_SearchResultsComponent_Host_0, {
  searchResults: "searchResults"
}

RenderType_FileNotFoundSearchComponent = createRendererType2({
  encapsulation: 2,
  styles: styles_FileNotFoundSearchComponent,
  data: {}
})
```

Is this expected? My understanding is that the host factories are only needed for entryComponents.

I can't estimate how significant this size regression is for aio because I'm unfortunately updating various dependencies. Due to bugs and api incompatibilities can't isolate the change just to the change of the compiler. Overall I'm looking for about 30kb size increase which might or might not contain other size regressions.

**Comment by maximusk** [comment]
I reported this behavior even for v4 [in this issue](https://github.com/angular/angular/issues/16984)

**Comment by erbsenkoenig** [comment]
Ich kehre zurück am 01.11.2017.

Ich werde Ihre Nachricht nach meiner Rückkehr beantworten.
In dringenden Fällen senden Sie bitte eine Kopie Ihrer E-Mail für
technische Angelegenheiten an entwicklung@arxes-tolina.de, ansonsten an
info@arxes-tolina.de. Ein anderer Mitarbeiter wird sich dann Ihrer E-Mail
annehmen.


Hinweis: Dies ist eine automatische Antwort auf Ihre Nachricht  "Re:
[angular/angular] size regression due to host view factories generated for
non-entryComponents (#19857)" gesendet am 23.10.2017 12:57:26.

Diese ist die einzige Benachrichtigung, die Sie empfangen werden, während
diese Person abwesend ist.

**Comment by IgorMinar** [comment]
@maximusk in my case the issue occurs with AOT compilation

**Comment by erbsenkoenig** [comment]
Ich kehre zurück am 01.11.2017.

Ich werde Ihre Nachricht nach meiner Rückkehr beantworten.
In dringenden Fällen senden Sie bitte eine Kopie Ihrer E-Mail für
technische Angelegenheiten an entwicklung@arxes-tolina.de, ansonsten an
info@arxes-tolina.de. Ein anderer Mitarbeiter wird sich dann Ihrer E-Mail
annehmen.


Hinweis: Dies ist eine automatische Antwort auf Ihre Nachricht  "Re:
[angular/angular] size regression due to host view factories generated for
non-entryComponents (#19857)" gesendet am 23.10.2017 13:05:11.

Diese ist die einzige Benachrichtigung, die Sie empfangen werden, während
diese Person abwesend ist.

**Comment by alexzuza** [comment]
I tried it with 4.3.6 and it also generated Host factories for non-entryComponents

Here's the piece of corresponding code:

https://github.com/angular/angular/blob/4.3.x/packages/compiler/src/aot/compiler.ts#L303

AFAIK only Compiler in JIT mode doesn't generate such factories for non-entryComponents

https://github.com/angular/angular/blob/4.3.x/packages/compiler/src/jit/compiler.ts#L174-L199

Jit compiler only generates host factories if:

- we call `compiler._compileModuleAndAllComponents`
- component is part of component's `entryComponents` metadata
- component is part of module's `entryComponents` metadata
- component is part of module's `bootstrap` metadata

So i think it was intended to generate host factories for all component in AOT

**Comment by tbosch** [comment]
@IgorMinar yes, AOT always used to generate host view factories for all components, as `entryComponents` could be introduced by a downstream dependency that is compiled in a separate compilation unit.

**Comment by tbosch** [comment]
The generated code here looks really wrong (it is after uglify and webpack though). The styles don't use the component factory, and the component factory should have been stored in a top level variable.

**Comment by tbosch** [comment]
@chuckjaz Steps to reproduce the incorrect code above (from @IgorMinar):

1. clone https://github.com/angular/angular/pull/19702
1. cd aio
1. yarn install (edited)
1. then you need to go and patch cli to disable uglify mangling
1. this is what I usually use:
```            // mangle: { screw_ie8: true },
            mangle: false,
            // compress: uglifyCompressOptions,
            compress: Object.assign(uglifyCompressOptions, {sequences: false}),
            // output: { ascii_only: true },
            output: { ascii_only: true, beautify: true },
            sourceMap: buildOptions.sourcemaps,
            comments: "all"
            // comments: false

```
1. it goes into aio/node_modules/@angular/cli/models/webpack-configs/production.js
1. find the uglify section and replace the config with the above
1. then build using:
1. ng build --target=production --environment=stable -sm --build-optimizer
1. and inspect less dist/main.*.bundle.js

**Comment by IgorMinar** [comment]
you can also disable the uglify completely by commenting it out in the webpack config. that will show you what we feed into uglify.

**Comment by chuckjaz** [comment]
The code above is incorrect. The `styles_FileNotFoundSearchComponent` should be `[]` and, if it had it correct value, would have been tree-shaken.

Removing uglify from the configuration, as described above by @IgorMinar, shows that the `ngc` is emitting the correct code. We also bashed the `writeFile` callback of the `compiler_host.js` in `ng` to save the intermediate output and the `ngc` output is correct.

It looks like the code gets mangled in the uglify step.

We tried executing uglifyjs directly, using several versions, but have not been able to duplicate the bug nor duplicate the output format so something might be going wrong in the uglify plugin or other code prior to uglify.

@hansl Is taking over the investigation to narrow down where the problem is in the webpack chain.

**Comment by hansl** [comment]
I just tried with the latest CLI master we are using the latest (released today) uglify-webpack-plugin 1.0.0 which includes UglifyES (which is the next generation Uglify).

The code above cannot be found. The `styles_` variables are properly inlined and only the factory call remains.

After the investigation and bisection of the problem with @chuckjaz, showing that:
1. removing `build-optimizer` did not fix the faulty optimization,
2. using `uglify-js` on a terminal did not reproduce the error,
3. removing uglify entirely, but keeping build optimizer on, fixed this issue,
4. doing a yarn list uglify-js shows 5 different installs, 2 of which are likely to clash (webpack and uglify-js), and
5. after an install of the CLI master we don't use the uglify-js package (instead relying on uglify-es),

I'm concluding that this is likely caused by a dependency problem with Uglify JS (between the root package and the one in webpack), that's fixed if you reinstall the CLI. I suggest removing it from the package.json in AIO, and updating the `yarn.lock`.

Please note that currently the size of AIO with all optimizations on, from the latest master of CLI (with no local patch), is:

```
$ ./node_modules/.bin/ng build --target=production --environment=stable --build-optimizer
Date: 2017-10-23T22:20:35.047Z
Hash: 5553ab53fc39c51b89ed
Time: 62687ms
chunk {0} 0.b19e913fbdd6507d346b.chunk.js () 14.8 kB {1}  [rendered]
chunk {1} main.2c656727784e224a18ec.bundle.js (main) 524 kB {4} [initial] [rendered]
chunk {2} polyfills.0dfca732c5a075c110d0.bundle.js (polyfills) 37 kB {4} [initial] [rendered]
chunk {3} styles.8991a9512e5bfea00ed0.bundle.css (styles) 115 kB {4} [initial] [rendered]
chunk {4} inline.326ec1e65b02b4cec54d.bundle.js (inline) 1.47 kB [entry] [rendered]
```

which is 27kb better than our baseline.

**Comment by IgorMinar** [comment]
@hansl removing uglify from the main package doesn't have significant impact on size, but it does change the output slightly. master is currently at 520kb, my upgrade branch is at 525kb and I was expecting 20-30kb in size savings from all the upgrades.

**Comment by IgorMinar** [comment]
I got to the bottom of this!!!

if I disable uglify I see this snippet of code in the bundle

```js
function View_SearchResultsComponent_Host_0(_l) { return viewDef(0, [(_l()(), elementDef(0, 0, null, null, 1, "aio-search-results", [], null, null, null, View_SearchResultsComponent_0, RenderType_SearchResultsComponent)), directiveDef(1, 573440, null, 0, search_results_component_SearchResultsComponent, [], null, null)], null, null); }

var SearchResultsComponentNgFactory = createComponentFactory("aio-search-results", search_results_component_SearchResultsComponent, View_SearchResultsComponent_Host_0, { searchResults: "searchResults" }, { resultSelected: "resultSelected" }, []);
```
The thing to notice is this line:
```
var SearchResultsComponentNgFactory = createComponentFactory("aio-search-results", 
```

The generated code has a top level function call `createComponentFactory` (and a few others) that are no longer being marked as @__PURE__, so uglify considers them as calls with potential side-effect and retains all of the code. (the code then gets inlined in various weird ways to decrease the size slightly, but it's essentially all retained.

In the past we used to have [special code in build-optimizer](https://github.com/angular/devkit/blob/7cfbb490d5b93597d340a6a7a26271a2457f5c3e/packages/angular_devkit/build_optimizer/src/purify/purify.ts#L56-L65) that would look for a few handpicked function calls in the generated code and mark them as pure. We however improved other bits and managed to inline core modules together with ng factories, which now means that we no longer can identify these calls via `angular_core` webpack module import, but those no longer exist, so we don't match these functions any more. This means that all of the generated code is fully retained even with build-optimizer turned on.

@filipesilva created a patch for build-optimizer that uses different approach to marking these functions as pure - we assume that all of the generated code has no non-local side-effects, so we prefix all the top level function calls with pure.

I tested this version of the optimization, and indeed all the generated code got marked as pure and removed by uglify if not retained by the rest of the app. This resulted in 35KB savings for aio.

@filipesilva is working on getting the patch into a release of build-optimizer now.

@tbosch @chuckjaz Now that we understand this, my question is: should we consider emitting the pure annotation directly from ngc for all generated code? That way people not using build-optimizer but using uglify would benefit.

I don't like generating uglify-specific instructions in our code, but it makes a huge impact on size and doesn't harm anything else. What do you think?

**Comment by filipesilva** [comment]
https://github.com/angular/devkit/pull/236 should address @IgorMinar's findings.

**Comment by hansl** [comment]
The fix in build-optimizer has been released 20 minutes ago.

**Comment by tbosch** [comment]
@IgorMinar If users are not using the build optimizer, a lot of other things are retained too, right?. So I would rather not add the dependency of our codegen to uglify specific comments but rather recommend our build optimizer in a stronger way.

**Comment by IgorMinar** [comment]
fixed by angular/devkit#236

**Comment by IgorMinar** [comment]
released as  @angular/devkit@0.0.31
